### DISCUSSION TITLE
Understanding pyuvm TLM interface classes

### DISCUSSION THREAD

**Discussion by jonpovey** [pyuvm/pyuvm#34] [Q&A] [answered]
I am reading `s12_uvm_tlm_interfaces.py` and trying to understand the relationship between port and export classes. I see the comments about exports being unnecessary, and that exports are all just empty subclasses of the corresponding ports.
I am trying to understand what this means for e.g. uvm_analysis_port and if there are unintended consequences.
As I read it, the consumer of analysis items is the `export` object passed to `connect()`, and `check_export()` is called to ensure the export is a subclass of `uvm_analysis_port`. I assume the expectation is that the export class overrides `write()` to actually make use of the analysis datum?
But it seems there are undesirable side effects from this class structure:

- The export gets a `connect()` method and `subscribers` attribute, `connect()` could be called and would work, but since `write()` is overridden, the upstream subscribers would get no items
- If `write()` is not overridden in the export subclass (or its name is a typo), the testbench will silently have incorrect behaviour, as `uvm_analysis_port.write()` called on the producer will run another copy of itself on the export object, which will silently immediately return as it has no subscribers. It seems preferable in this case that `export.write()` should not exist and raise an `AttributeError`, i.e. the export object should not inherit from `uvm_analysis_port`.

I think that the other port types may have similar problems, but I am not experienced with UVM and perhaps I am missing something. Clues appreciated, thanks.

**Comment by raysalemi** [comment, marked as answer]
> I am reading `s12_uvm_tlm_interfaces.py` and trying to understand the relationship between port and export classes. I see the comments about exports being unnecessary, and that exports are all just empty subclasses of the corresponding ports. I am trying to understand what this means for e.g. uvm_analysis_port and if there are unintended consequences. As I read it, the consumer of analysis items is the `export` object passed to `connect()`, and `check_export()` is called to ensure the export is a subclass of `uvm_analysis_port`. I assume the expectation is that the export class overrides `write()` to actually make use of the analysis datum? But it seems there are undesirable side effects from this class structure:
> 
> * The export gets a `connect()` method and `subscribers` attribute, `connect()` could be called and would work, but since `write()` is overridden, the upstream subscribers would get no items
> * If `write()` is not overridden in the export subclass (or its name is a typo), the testbench will silently have incorrect behaviour, as `uvm_analysis_port.write()` called on the producer will run another copy of itself on the export object, which will silently immediately return as it has no subscribers. It seems preferable in this case that `export.write()` should not exist and raise an `AttributeError`, i.e. the export object should not inherit from `uvm_analysis_port`.
> 
> I think that the other port types may have similar problems, but I am not experienced with UVM and perhaps I am missing something. Clues appreciated, thanks.

You're right!  I never saw it that way. 

To paraphrase you're pointing out that export classes don't have a connect function, but extending port to create them incorrectly gives them a connect function.  

I think the answer is to make `uvm_analysis_export` the base class and extend it to create `uvm_analysis_port` that defines a connect function.  That way the connect function could check that it is receiving an export, and this will be true with a port or export, while not allowing incorrect calls to connect from an export.

Good catch!  I'll create an issue so I don't forget.

**Reply by jonpovey** [reply]
That's interesting, I had thought that uvm_analysis_export should be a separate class to uvm_analysis_port. But like I said, I am still trying to get my head around the port/export/impl design.

**Reply by raysalemi** [reply]
> That's interesting, I had thought that uvm_analysis_export should be a separate class to uvm_analysis_port. But like I said, I am still trying to get my head around the port/export/impl design.

You need to be able to connect a port to a port. In the end you need a connection to an export class that implements the `get()`, `put()`, `write()` or whatever the port needs.

**Reply by jonpovey** [reply]
Right, is this because you want to do some type-checking, maybe ensure some code in `uvm_port_base` will get run?
AIUI the Pythonic duck-typing way would be to say the 'export' object can be any type; so long as it implements a `write(self, datum)` method, everything is fine. Otherwise a runtime exception will be raised. (Or `connect` could check at that time that `export.write()` exists)

**Reply by raysalemi** [reply]
Checking for write() is a better solution. I like it.

**Reply by jonpovey** [reply]
Glad that was helpful. Due to my lack of experience with SV/UVM, I don't know how much of pyuvm is "it needs to be this way to be like UVM" and how much is open to change. Such as all the interface subclasses in `uvm_tlm_fifo_base` that look to me like they could just be methods... But that is a different topic.

**Comment by raysalemi** [comment]
There’s a tension between implementing the specification and being
Pythonic.
